addons/sale_purchase/models/sale_order_line.py

```diff
@@ -159,99 +190,81 @@ def _purchase_service_prepare_line_values(self, purchase_order, quantity=False):
-            supplierinfo = suppliers[0]
+            line = line.with_company(line._purchase_service_get_company())
+            supplierinfo = self._purchase_service_match_supplier()
             partner_supplier = supplierinfo.partner_id
 
             # determine (or create) PO
             purchase_order = supplier_po_map.get(partner_supplier.id)
             if not purchase_order:
-                purchase_order = PurchaseOrder.search([
-                    ('partner_id', '=', partner_supplier.id),
-                    ('state', '=', 'draft'),
-                    ('company_id', '=', line.company_id.id),
-                ], limit=1)
-            if not purchase_order:
-                values = line._purchase_service_prepare_order_values(supplierinfo)
-                purchase_order = PurchaseOrder.with_context(mail_create_nosubscribe=True).create(values)
-            else:  # update origin of existing PO
+                purchase_order = self._match_or_create_purchase_order(supplierinfo)
+            else:  # if not already updated origin in this loop
                 so_name = line.order_id.name
-                origins = []
-                if purchase_order.origin:
-                    origins = purchase_order.origin.split(', ') + origins
+                origins = (purchase_order.origin or '').split(', ')
                 if so_name not in origins:
-                    origins += [so_name]
-                    purchase_order.write({
-                        'origin': ', '.join(origins)
-                    })
-            supplier_po_map[partner_supplier.id] = purchase_order
+                    purchase_order.write({'origin': ', '.join(origins + [so_name])})
+                supplier_po_map[partner_supplier.id] = purchase_order
 
             # add a PO line to the PO
             values = line._purchase_service_prepare_line_values(purchase_order, quantity=quantity)
-            purchase_line = line.env['purchase.order.line'].create(values)
+            purchase_lines |= line.env['purchase.order.line'].create(values)
 
-            # link the generated purchase to the SO line
-            sale_line_purchase_map.setdefault(line, line.env['purchase.order.line'])
-            sale_line_purchase_map[line] |= purchase_line
-        return sale_line_purchase_map
+        return purchase_lines
 
     def _purchase_service_generation(self):
         """ Create a Purchase for the first time from the sale line. If the SO line already created a PO, it
             will not create a second one.
         """
-        sale_line_purchase_map = {}
         for line in self:
             # Do not regenerate PO line if the SO line has already created one in the past (SO cancel/reconfirmation case)
             if line.product_id.service_to_purchase and not line.purchase_line_count:
-                result = line._purchase_service_create()
-                sale_line_purchase_map.update(result)
-        return sale_line_purchase_map
```

Comment:
We should not change the method signature in a frozen version.

Reply:
reverted to minimal diff, but a lot of useless code...